### DISCUSSION TITLE
Disabling the apheleia indent step

### DISCUSSION THREAD

**Discussion by ispringle** [radian-software/apheleia#208] [Q&A] [answered]
I might not be understanding the purpose of the apheleia indent step but I think it's resulting in code that's not properly formatted. I'm trying to format typescript/javascript on save and what is returned by apheleia is not what prettier suggests. I was confused for a bit why there is consistently an extra 4 spaces returned by apheleia when prettier is using an indent level of two spaced. I looked in the codebase here and it seems apheleia pipes most/all code through the given formatter and then pipes that through its own indentation function.

Trying to understand the logic and point of overriding the formatter here and either working with apheleia or just ripping this indentation out of all the configurations.

**Comment by raxod502** [comment]
Where are you seeing that Apheleia is performing an additional indentation step besides what the formatter suggests?

**Reply by ispringle** [reply]
My apologies. I re-read the default prettier formatting and it's actually just supplying prettier with an indent leven based on the js-indent variable. https://github.com/radian-software/apheleia/blob/e944e24584393c1a4df130885b489795f7676567/apheleia.el#L95

I think I will have to rewrite all of these to let prettier manage indent level, unless there is another way to just allow the default prettier config do its thing without being given the indent levels.

**Reply by raxod502** [reply]
Well, the intent of the indentation configuration is to have the formatter match how Emacs is configured to format things. Because otherwise you'd get stuff indented wrong as you type, before being fixed on save.

We might as well make this configurable though. It is reasonable to want the Emacs settings to be superseded by your formatter settings.

**Reply by raxod502** [reply, marked as answer]
Give https://github.com/radian-software/apheleia/pull/211 a try.

**Reply by ispringle** [reply]
Thanks for the work there! This works great!

**Reply by mohkale** [reply]
I kind of disagree with this choice, if your editing experience in Emacs doesn't match the end result of your formatter you should configure emacs correctly rather than overwrite your editor with the formatter. I think something like tabs or tab-width should be consistently configured otherwise it'll probably just be a constant annoyance until you save. I know when I'm in a json-mode buffer configured to indent with 2 spaces when I default to 4 that it annoys me when I enter a newline and suddenly the indentation is off.

@ispringle is there a reason you don't want to configure emacs to indent the same way as your formatter does? Generally you can do this for individual projects if you lookup which variable sets the indentation width and style. Also using editorconfig is generally recommended to avoid issues like this.

**Reply by ispringle** [reply]
We use prettier at work as a test/check for PRs. If a PR doesn't conform to prettier (ie prettier detects discrepancies) I can't merge. But I don't like what prettier looks like either, I find that 2 spaces isn't enough for me to differentiate blocks/levels well. When I'm working on my project I want it to be four spaces but when I go to save a file prettier needs to be the source of truth, not emacs. 

With this I just can toggle off the emacs indent, let prettier do it's thing, commit, and then toggle emacs' indent levels back on and have readable code.

**Reply by mohkale** [reply]
Why not run prettier in a git-commit hook then and avoid formatting at all until you definitively have to? It seems like you're trying to avoid formatting instead of trying to standardize with a formatter. Perhaps that's a valid use case although I find it no less bizarre :thinking:.

**Reply by mohkale** [reply]
For reference I do something similar, at my work we use clang-format but I haven't setup a code-style conforming to it so I just don't run apheleia their at all given my editing experience would always be out of sync with what happens on save. There's a test which validates code written against a formatter and I just run it before opening PRs. If it fails just batch clang-format. I should probably setup a git hook but just haven't had the need yet.

**Reply by ispringle** [reply]
The main repo I work on does that, and for that project there is no issue. I've actually been using Apheleia for some time with no problems and no need for this. I didn't even know Apheleia did this since what Apheleia spits out and what this project's Prettier spit out are identical. However I've been recruited to work on some other projects at work and they do not have the same setup and they're not interested in changing their workflow. When I mentioned pre-commit hooks in the past you'd have thought I told them I wanted their first born child. They're likewise not a big fan of a few other things I've tried, including "polluting" their project with various configuration files.

Anyway, that's when I first realized Apheleia was actually even running a post formatter indentation. I think the flexibility to turn it off is nice, helps with weird work flows.

**Reply by mohkale** [reply]
You don't need to commit a pre-commit hook into a repo to use it, you can have a project specific pre-commit hook just for your local clone of the repo. There's the question of whether you should or not but if everyone else is resistant to it I see no harm in just customizing a repos configuration for your own developments. You could also just not use the built-in formatter definition. Add a .dir-locals-2.el file to the repo, git ignore it and set it up to assign `apheleia-formatters`  so the the specific formatter you're using doesn't set the flags for the indentation. There's a couple different ways to crack this egg from my POV, I disagree with doing so here though. Generally I dislike packages with too many options catering to different workflows, I prefer when you make things flexible enough that people can fix problems themselves and this would fit that case more IMO :smile:.

Edit: Perhaps I'm mistaken, looks like a couple other folks are confused by this as well #189. Not really sure what the best solution here would be. Maybe for formatters that depend on project specific configurations like prettier we don't set any indent options?

**Reply by raxod502** [reply]
I don't think there's any correct answer. It will have to be configurable because there are some cases where it is unambiguously correct to follow how Emacs is configured, and other cases where it is unambiguously correct to follow how the formatter wants to default things. It won't be possible to distinguish between these in a fully automatic way in all cases.

> that's when I first realized Apheleia was actually even running a post formatter indentation

For the record, this is not what's happening. There is no post-formatter modification at all. It is just that Apheleia runs the formatter with options based on how you've configured Emacs to indent things.

> toggle emacs' indent levels back on

What does that mean? Are you using tabs and changing the visual tab width? Otherwise, I'm not sure what "turning indentation back on" would even mean, it is not as if enabling e.g. `electric-indent-mode` suddenly rewrites all the lines in your buffer to have a different number of leading spaces.

**Reply by ispringle** [reply]
re: toggle

I meant toggling t/nil of the variable you added.

**Reply by raxod502** [reply]
Oh, I see. Ok, that makes sense.
